compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp

```diff
@@ -17,6 +19,271 @@ namespace mlir::iree_compiler {
 
 namespace {
 
+/// Calculate the expanded shape of `dest` if it can be expanded with the inner
+/// expanded sizes of `sliceStaticSizes`. Returns failure if such expansion is
+/// not possible.
+static LogicalResult
+getExpandedShape(SmallVector<ReassociationIndices> reIndices,
+                 ArrayRef<int64_t> sliceStaticSizes, Value dest,
+                 SmallVector<int64_t> &expandedShape,
+                 SmallVector<int64_t> &totalInnerSizes) {
+  auto destType = dyn_cast<ShapedType>(dest.getType());
+  if (!destType)
+    return failure();
+  // Iterator to insert outer sizes.
+  auto outerShapeIter = expandedShape.begin();
+  for (auto [reassociations, destSize] :
+       llvm::zip_equal(reIndices, destType.getShape())) {
+    int64_t totalInnerSize = 1;
+    for (int i = 1; i < reassociations.size(); i++) {
```

Comment:
do not recalculate range end and prefer pre-increment: https://llvm.org/docs/CodingStandards.html

Reply:
I am not sure I follow the `do not recalculate range end ` part. I dont have the value of reassociations.size() anywhere else..

Reply:
the canonical llvm loop form is `for (size_t i = 1, e = reassociations.size(); i != e; ++i)`

Reply:
https://github.com/iree-org/iree/pull/19044